### PR TITLE
feature: Support ReactiveUI WhenActivated

### PR DESCRIPTION
**Pull request by worldbeater**
Heavily based on https://github.com/reactiveui/ReactiveUI/pull/1748
See https://github.com/reactiveui/ReactiveUI/pull/1748#issuecomment-427612442

**What does the pull request do?**

This PR adds ReactiveUI <a href="https://reactiveui.net/docs/handbook/when-activated/">WhenActivated</a> and type-safe bindings support for AvaloniaUI views.

> WhenActivated is a way to track disposables. Besides that, it can be used to defer the setup of a ViewModel until it's truly required. WhenActivated also gives us an ability to start or stop reacting to hot observables, like a background task that periodically pings a network endpoint or an observable updating users current location. — <a href="https://reactiveui.net/docs/handbook/when-activated/">ReactiveUI Docs</a>

**What is the current behavior?**

Currently, I get an exception when I try to use `WhenActivated` feature with Avalonia.

**What is the updated/expected behavior with this PR?**

<a href="https://reactiveui.net/docs/handbook/when-activated/">WhenActivated</a> will work with AvaloniaUI as expected. See https://github.com/AvaloniaUI/Avalonia/issues/1852

**How was the solution implemented?**

I've implemented and registered an interface to let ReactiveUI know abount the lifecycle of `IVisual`. 

**Checklist:**

- [x] Add `IActivationForViewFetcher` implementation
- [x] Register `IActivationForViewFetcher` implementation
- [x] Add unit tests (brought some from https://github.com/reactiveui/ReactiveUI/pull/1748)

### CODE REVIEW

**Comment by worldbeater** [comment]
The PR is almost ready to ship, but needs a review. Now Avalonia supports <a href="https://reactiveui.net/docs/handbook/when-activated/">WhenActivated</a> — a ReactiveUI way of tracking subscriptions and lifecycles of reactive ViewModels.

There is something wrong with CI, but I doubt this is my fault:
```
Error: One or more errors occurred.
Error downloading 'NuGet.CommandLine.4.3.0' from 
'https://api.nuget.org/v3-flatcontainer/nuget.commandline/4.3.0/nuget.commandline.4.3.0.nupkg'.
```

File: tests/Avalonia.ReactiveUI.UnitTests/AvaloniaActivationForViewFetcherTest.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+using Xunit;
+using Splat;
+
+namespace Avalonia 
+{
+    public class AvaloniaActivationForViewFetcherTest
+    {
+        public class TestUserControl : UserControl, IActivatable { }
+
+        public class FakeRenderDecorator : Decorator, IRenderRoot
```
Comment (kekekeks) [inline tests/Avalonia.ReactiveUI.UnitTests/AvaloniaActivationForViewFetcherTest.cs]: Use [TestRoot](https://github.com/AvaloniaUI/Avalonia/blob/master/tests/Avalonia.UnitTests/TestRoot.cs) instead.

File: tests/Avalonia.ReactiveUI.UnitTests/Avalonia.ReactiveUI.UnitTests.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (kekekeks) [inline tests/Avalonia.ReactiveUI.UnitTests/Avalonia.ReactiveUI.UnitTests.csproj]: You need to add this project [here](https://github.com/AvaloniaUI/Avalonia/blob/master/build.cake#L172)

**Review by kekekeks — DISMISSED** [review]

File: samples/VirtualizationDemo/ViewModels/MainWindowViewModel.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 using Avalonia.Collections;
 using Avalonia.Controls;
 using Avalonia.Controls.Primitives;
+using ReactiveUI.Legacy;
```
[the remark this answers is not in the document]
Comment (worldbeater) [reply samples/VirtualizationDemo/ViewModels/MainWindowViewModel.cs]: ReactiveList is used there, which is no longer a recommended way to work with collections, is marked obsolete and moved to ReactiveUI.Legacy namespace. 

We should consider rewriting that sample using DynamicData now or later. See https://reactiveui.net/blog/2018/10/reactiveui-v9.0.1-released

Here, I use ReactiveUI.Legacy namespace to make the project compile.

**Comment by grokys** [comment]
Apologies for the delay in reviewing this! How do I apply `WhenActivated` to a view? `this.WhenActivated` doesn't compile.

**Comment by worldbeater** [comment]
@grokys you need to implement either `IViewFor<TViewModel>` or `IActivatable` marker interface on your UserControl. See <a href="https://github.com/reactiveui/ReactiveUI/blob/master/samples/getting-started/ReactiveDemo/MainWindow.xaml.cs#L7">a simple example</a>, or a more complex one: <a href="https://github.com/worldbeater/MyFeed/blob/master/MyFeed.Uwp/Views/FeedGroupView.xaml.cs#L9">activatable view</a>, <a href="https://github.com/worldbeater/MyFeed/blob/master/MyFeed/ViewModels/FeedGroupViewModel.cs#L19">activatable viewmodel</a>. Also, we might want to create a <a href="https://reactiveui.net/docs/handbook/when-activated/#view">ReactiveUserControl<TViewModel></a> for Avalonia, that will encapsulate `IViewFor<TViewModel>` implementation (<a href="https://github.com/reactiveui/ReactiveUI/blob/a9e3373962c35daf8cea0129ae2b5856f4f0a5e5/src/ReactiveUI/Platforms/windows-common/ReactiveUserControl.cs#L43">like this one does</a>). If we don't have a ViewModel, marking a UserControl with `IActivatable` interface is enough (I used this approach in <a href="https://github.com/AvaloniaUI/Avalonia/pull/1951/files#diff-a6e58f304515d0dfe99baf37bfc6ed51R20">tests</a>). Thanks for reviewing this!

**Comment by grokys** [comment]
Ah yes, I did actually know that at some point, thanks @worldbeater! Giving it a try now.

**Review by grokys — APPROVED** [review]
This is great, thanks @worldbeater !

**Pull request merged**